File: problems/scaler/Two Pointers/Area_containing_max_water.js
/**
 * Given an array of integers A of size N, where A[i] represents the height of the ith wall. Find the maximum amount of water that can be trapped in the walls.
 *
 * The amount of water that can be trapped in the walls is the difference between the height of the wall and the height of the water level.
 *
 * The water level is the minimum of the height of the wall and the height of the water level.
 *
 */

/**
 * Approach: using two pointers
 */
function soln(A) {
  const n = A.length;
  let i = 0;
  let j = n - 1;
  let maxWater = 0;
  while (i < j) {
    maxWater = Math.max(maxWater, Math.min(A[i], A[j]) * (j - i));
    if (A[i] < A[j]) {
      i++;
    } else {
      j--;
    }
  }
  return maxWater;
}

console.log(soln([1, 8, 6, 2, 5, 4, 8, 3, 7])); // 49

File: src/main/java/com/dsa/problems/two_pointers/Area_Containing_Max_Water.java
/**
 * Given an array of integers A of size N, where A[i] represents the height of the ith wall. Find the maximum amount of water that can be trapped in the walls.
 *
 * The amount of water that can be trapped in the walls is the difference between the height of the wall and the height of the water level.
 *
 * The water level is the minimum of the height of the wall and the height of the water level.
 *
 */
package com.dsa.problems.two_pointers;

import java.util.ArrayList;
import java.util.Arrays;

public class Area_Containing_Max_Water {
  public static int maxWaterTrapped(ArrayList<Integer> A) {
    int n = A.size();

    int i = 0;
    int j = n - 1;

    int maxWater = 0;

    while (i < j) {
      int water = Math.min(A.get(i), A.get(j)) * (j - 1);
      maxWater = Math.max(maxWater, water);

      if (A.get(i) < A.get(j)) {
        i++;
      } else {
        j--;
      }
    }

    return maxWater;
  }

  public static void main(String[] args) {
    System.out.println(maxWaterTrapped(new ArrayList<Integer>(Arrays.asList(1, 5, 4, 3))));
  }
}
